ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_device_operation.cpp

```diff
@@ -0,0 +1,331 @@
+
+    Tensors input_tensors = {tensor_args.a, tensor_args.b};
+    if (tensor_args.bias.has_value()) {
+        input_tensors.push_back(tensor_args.bias.value());
+    }
+
+    tt::tt_metal::operation::OpPerformanceModelGeneral<tensor_return_value_t> result(
+        input_tensors, output_tensor, ideal_dev_clock_cycles);
+
+#if 0
+    log_info(tt::LogOp, "Conv2d PerfModel:");
+    log_info(tt::LogOp, "\t Batch: {}", batch_size);
+    log_info(tt::LogOp, "\t In (H, W, C): ({}, {}, {})", conv_activation_h, conv_activation_w, conv_activation_c);
+    log_info(tt::LogOp, "\t Filter (H, W): ({}, {})", filter_h, filter_w);
+    log_info(tt::LogOp, "\t Filter Stride (H, W): ({}, {})", stride_h, stride_w);
+    log_info(tt::LogOp, "\t Pad (H, W): ({}, {})", pad_h, pad_w);
+    log_info(tt::LogOp, "\t Out (H, W, C): ({}, {}, {})", output_height, output_width, this->output_channels);
+    log_info(tt::LogOp, "\t ideal_dev_clock_cycles: {}", ideal_dev_clock_cycles);
+#endif
+
+    return result;
+}
+
+std::tuple<operation_attributes_t, tensor_args_t> Conv2dDeviceOperation::invoke(
+    const Tensor& a,
+    const Tensor& b,
+    std::optional<const Tensor> bias,
+    const sliding_window::SlidingWindowConfig& sliding_window_config,
+    uint32_t output_channels,
+    uint32_t groups,
+    bool untilize_out,
+    bool has_bias,
+    const std::optional<ttnn::operations::unary::UnaryWithParam>& activation,
+    const Conv2dParallelizationConfig& parallelization_config,
+    const Conv2dBlockConfig& block_config,
+    const tt::tt_metal::MemoryConfig& memory_config,
+    tt::tt_metal::DataType dtype,
+    std::array<std::uint32_t, 4> input_tensor_shape,
+    const DeviceComputeKernelConfig& compute_kernel_config,
+    bool enable_act_double_buffer,
+    bool enable_weights_double_buffer,
+    bool full_inner_dim,
+    bool enable_activation_reuse,
+    bool config_tensors_in_dram,
+    std::optional<bool> force_split_reader) {
+    auto args = operation_attributes_t{
+        .sliding_window_config = sliding_window_config,
+        .output_channels = output_channels,
+        .groups = groups,
+        .untilize_out = untilize_out,
+        .has_bias = has_bias,
+        .activation = activation,
+        .parallelization_config = parallelization_config,
+        .block_config = block_config,
+        .memory_config = memory_config,
+        .dtype = dtype,
+        .input_tensor_shape = input_tensor_shape,
+        .compute_kernel_config = compute_kernel_config,
+        .enable_act_double_buffer = enable_act_double_buffer,
+        .enable_weights_double_buffer = enable_weights_double_buffer,
+        .full_inner_dim = full_inner_dim,
+        .enable_activation_reuse = enable_activation_reuse,
+        .config_tensors_in_dram = config_tensors_in_dram,
+        .force_split_reader = force_split_reader,
+
+    };
+    auto tensor_args = tensor_args_t{
+        .a = a,
+        .b = b,
+        .bias = bias,
+    };
+
+    IDevice* device = a.device();
```

Comment:
lets use auto here. idevice might go away
```suggestion
    auto device = a.device();
```

Reply:
Changed all variables holding a `.device()` return value to type `auto` in [2b90315](https://github.com/tenstorrent/tt-metal/pull/33019/commits/2b903150997a85024b117979ae9078b1672a5cd9).